### PR TITLE
Feat: added use_latest_tip query parameter to the relevant v2 endpoints

### PR DESCRIPTION
**Pull request by pavitthrap**
## Description
Updated `tip` query parameter to the v2 RPC endpoints to accept the value of "latest". 

When set to "latest", the latest tip is used - this includes the unconfirmed state, if there is any unconfirmed state. 

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] API reference/documentation update
- [ ] Other

## Are documentation updates required?
Added changes to the openapi yaml file 

## Testing information
Added tests mostly in `rpc.rs`. 
Added test in `integrations.rs`.

### CODE REVIEW

File: src/net/rpc.rs
Comment (gregorycoppola) [inline src/net/rpc.rs:3146]: does it make sense to test this new code? Is it possible? Do you think it's self-testing?

File: src/net/rpc.rs
Comment (gregorycoppola) [inline src/net/rpc.rs:4393]: Can we have a few comments about why we expect the answers in the tests to be the way they are? And why there are not expected to be what they are not?

**Review by gregorycoppola — COMMENTED** [review]
Hello. Seems like a good change but I just had a few nits about the comments, including for the tests to add some info about the logic of the test for people new to the file.

File: src/net/rpc.rs
Comment (pavitthrap) [reply src/net/rpc.rs:3146]: The code itself is a test, so I'm not sure - also this is not new code, I added an indented section to make including microblocks optional.

File: src/net/rpc.rs
Comment (pavitthrap) [reply src/net/rpc.rs:4393]: Added comment.

File: docs/rpc/openapi.yaml
Comment (zone117x) [inline docs/rpc/openapi.yaml:437]: Thoughts on making this more OpenAPI-compliant with a `boolean` rather than a `string`? It helps with doc and client generation. 

In the API, I usually document these as `boolean` (`true`/`false`), but in the route handling code also support parsing `1`/`0`.

Comment (pavitthrap) [reply docs/rpc/openapi.yaml:437]: sure, I'll make this change

File: src/net/rpc.rs
Comment (gregorycoppola) [reply src/net/rpc.rs:4405]: this is a very helpful comment!, but can you please do the same for the other tests you are adding?

**Review by gregorycoppola — COMMENTED** [review]
Looks good.. can you just add comments for all tests you are adding?

File: docs/rpc/openapi.yaml
Comment (lgalabru) [inline docs/rpc/openapi.yaml:82]:
```suggestion
          If use_latest_tip == true, the query will be run from the latest known tip (includes unconfirmed state).
```

File: src/net/rpc.rs
Comment (gregorycoppola) [reply src/net/rpc.rs:3159]: ok

File: src/net/rpc.rs
Comment (gregorycoppola) [inline src/net/rpc.rs:3402]: so it defaults to `use_latest_tip` is true? If so please update comment

File: src/net/rpc.rs
Comment (gregorycoppola) [inline src/net/rpc.rs:4236]: "only exists"

File: src/net/rpc.rs
Comment (gregorycoppola) [reply src/net/rpc.rs:4476]: nice comments, thank you!

**Review by gregorycoppola — APPROVED** [review]
Thanks for all the new test comments.. I feel like I can understand it a lot better and hopefully will be the same for other readers as well!

File: src/net/rpc.rs
Comment (gregorycoppola) [inline src/net/rpc.rs:1524]: This is kind of confusing, how about "Otherwise, the function considers the canonical Stacks chain tip."

**Review by gregorycoppola — APPROVED** [review]
Thanks for the change! One small nit on the comments but LGTM.

File: src/net/http.rs
Comment (jcnelson) [inline src/net/http.rs:1684]: Do we gain anything concrete by accepting either/or `1` or `true`?  Might as well just pick one or the other if not.

File: src/net/http.rs
Comment (jcnelson) [inline src/net/http.rs:2567]: You don't have to address it here, but could you open an issue to consolidate `tip_opt`, `with_proof`, and `use_latest_tip` into a single struct that contains the standard query parameters that most API endpoints accept?  Otherwise PRs like this are going to get untenable ;)

Comment (pavitthrap) [reply src/net/http.rs:2567]: sure!

File: src/net/http.rs
Comment (zone117x) [reply src/net/http.rs:1684]: Yep, if we want to adhere to OpenAPI schema and general API design guidelines, this should be a boolean rather than an `number` or `string`. I suggested also allowing a number for both consistency purposes (some existing endpoints use a number rather than a boolean), as well as the robustness principal (permissive in what is accepted) because this isn't really a security-sensitive feature. 

In the stacks-blockchain-api, similar boolean query params are pretty liberal in what is accepted, because it's not entirely unreasonable for someone to do any of the following:
`curl http://example.com/?use_latest_tip=true`,
`curl http://example.com/?use_latest_tip=1`, or even
`curl http://example.com/?use_latest_tip`. 

But, if there are reservations around accepting multiple forms of query param input, it should be a proper boolean.

Comment (jcnelson) [reply src/net/http.rs:1684]: What's here is fine.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:4240]: Please be sure to change the test name and the ports used here.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:4333]: Please be sure to change the test name and ports here.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:4416]: Please be sure to change the test name and ports here.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:4674]: Same here please.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:4814]: Same here

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:4901]: Same here

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:1549]: Please don't `.unwrap()` in network-facing code.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:1535]: I just want to confirm that the following scenarios are tested:

* The caller passes `use_latest_tip=1` with an existing unconfirmed microblock stream, and Clarity state from the unconfirmed microblock stream is successfully loaded.
* The caller passes `use_latest_tip=1` with an empty unconfirmed microblock stream, and Clarity state from the canonical chain tip is successfully loaded (i.e. you don't get a 404 since the unconfirmed chain tip points to a nonexistent MARF trie).
* The caller passes `use_latest_tip=0`, and the canonical chain tip is used regardless of the state of the unconfirmed microblock stream.

Comment (jcnelson) [reply src/net/rpc.rs:1535]: That second bullet is going to be a bit thorny.  It may work in these unit tests, but the unit test framework doesn't reflect how the Stacks node actually deals with unconfirmed chain state.  This is because the Stacks node itself has a very weird way of representing unconfirmed state, due to the fact that unconfirmed state is curated both by the p2p thread (to service read-only queries off of it) and by the relayer thread (to process new microblocks).  I think you might want to create a nominal integration test for these bullet points in `neon_integrations.rs`.  It doesn't have to exhaustively test all of the RPC methods; it just needs to verify that the right chain tip is used in each situation (and nothing 404's).

**Review by jcnelson — CHANGES_REQUESTED** [review]
Overall this looks good to me @pavitthrap!  I just had a couple of small but must-have testing asks.  After that, I'll approve.

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:1549]: @pavitthrap This issue isn't resolved yet, so I'm unresolving the conversation.

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:4240]: Unresolving this conversation, since the issue is not resolved yet.

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:4333]: Unresolving this conversation since the issue is not yet resolved.

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:4416]: Unresolving this conversation since the issue is not yet resolved.

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:4674]: Unresolving

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:4814]: Unresolving

File: src/net/rpc.rs
Comment (jcnelson) [reply src/net/rpc.rs:4901]: Unresolving

**Comment by jcnelson** [comment]
Review is blocked on #2774

File: src/net/rpc.rs
Comment (pavitthrap) [reply src/net/rpc.rs]: Addressed in commit 0c5f6aa

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:1586]: This API feels a little weird.  According to this code path, if I requested the query string `?tip=xxxxxxx&use_latest_tip=true`, not only would the request potentially succeed, but also the `tip=` argument would be ignored in favor of `use_latest_tip`.  Is this what we want?  Would it make more sense to require the caller to use *either* `tip=` *or* `use_latest_tip` but *not* both?

Comment (jcnelson) [reply src/net/rpc.rs:1586]: It makes me wonder if instead of having a separate `use_latest_tip` query argument, we could instead overload `tip=` so that `tip=latest` becomes a valid argument.  Then, this ambiguity goes away.

Comment (pavitthrap) [reply src/net/rpc.rs:1586]: I think I like adding the option for `tip=latest`. I'll refactor this PR to do that.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:1550]: Minor refactoring nit: can you add a method to `UnconfirmedState` called `get_unconfirmed_tip_if_exists()` that does all this work here?  Then, if it fails -- i.e. the underlying trie doesn't exist -- this code block would simply return the appropriate HTTP response type.

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:1561]: This wouldn't be a server error (500).  It would be a not-found error (404).

File: src/net/rpc.rs
Comment (jcnelson) [inline src/net/rpc.rs:1592]: This would also be a 404, not a 500.  A stacks node that hasn't downloaded any blocks yet is not in an error state.

File: testnet/stacks-node/src/tests/neon_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/neon_integrations.rs:681]: I think this PR got rid of `use_latest_tip`, and now just uses `tip=latest`?

File: docs/rpc/openapi.yaml
Comment (jcnelson) [inline docs/rpc/openapi.yaml:77]: I think this PR changes this from `use_latest_tip` to `tip=latest`?

**Comment by jcnelson** [comment]
This PR is looking a lot better.  Thanks for adding all the thorough test coverage!  Just a few minor nits to address, and then I'll approve.

**Review by jcnelson — APPROVED** [review]
LGTM.  I think the one thing that still stands out to me on this PR is that there's a race condition whereby the unconfirmed tip obtained from `handle_load_stacks_chain_tip()` can get invalidated before getting passed into `maybe_read_only_clarity_tx()`.  However, the result of this race occurring isn't too bad -- the caller would simply receive a 404 response.  I don't think this is a show-stopping problem; if it becomes a problem, the fix would be to simply load the unconfirmed chain tip hash again at the call for `maybe_read_only_clarity_tx()` if it got invalidated.  Can you add a TODO comment describing this on `handle_load_stacks_chain_tip()` before merging (and open an issue for it)?

**Pull request merged**
